### PR TITLE
fix: ensure pubkey is always present on AWS

### PR DESCRIPTION
**Pull request by ezr-ondrej**
Ensure pubkey is on AWS and save its name to ASWDetails.

Fixes: HMSPROV-339 HMSPROV-332

Refactors the pubkey job to Ensure the pubkey exists on AWS.
First it tries to look for the key on AWS, if it exits, it gets its name.
If it doesn't exist, it imports the key with the name we have in our DB.

This name gets saved to AWS details so next job can fetch it and pass the correct name to AWS `RunInstances`.
Adds tests for the refactored job.

### CODE REVIEW

**Review by lzap — CHANGES_REQUESTED** [review]
Few nitpick.

Tho I am not sure if we should tag existing resources, in the future we might be able to delete them and this might be uploaded extenrally.

But for keypair it is probably okay.

File: internal/clients/http/ec2/ec2_client.go
Comment (lzap) [reply internal/clients/http/ec2/ec2_client.go:171]: Ok now you have shadowing, I know this one is pesky but shadowing can introduce nasty bugs.

**Comment by ezr-ondrej** [comment]
Well you might be actually right :thinking: maybe we should just blank the tag in the database and that would mean this is not a resource we've created :thinking:

**Comment by lzap** [comment]
Commit message check complains, feel free to update it if you want to use tickets like this.

**Comment by lzap** [comment]
Tho I would rather prefer dropping them directly in the commit message, we can easily parse them all since they have a common prefix after changelog generation via `perl` or `awk`.

**Comment by ezr-ondrej** [comment]
I'm updating it in https://github.com/RHEnVision/provisioning-backend/pull/313/files

**Comment by lzap** [comment]
Please rebase and reword the commit message depending on the outcome the team has on the subject.

Or you can just reword this in the old way for now so this can be merged.

File: internal/clients/http/ec2/ec2_client.go
Comment (adiabramovitch) [inline internal/clients/http/ec2/ec2_client.go:203]: I think it will be more readable to save *output.KeyPairs[0] as a variable

File: internal/jobs/pubkey_upload_aws_job.go
Comment (ezr-ondrej) [inline internal/jobs/pubkey_upload_aws_job.go:91]: :warning: Handle would be a name of the key now. Can we handle it? We will need to do something about our DB values on stage (cleanup of the Pubkey resources comes to mind as simple solution).

Reasoning: we are passing a name to the `RunInstances`, so it should be the handle, what for we keep the Handle if not to be used to be passed to `RunInstances`.
@lzap @adiabramovitch

File: internal/clients/http/ec2/ec2_client.go
Comment (ezr-ondrej) [reply internal/clients/http/ec2/ec2_client.go]: Good suggestion, but I've refactored the code completely, so now it doesn't make sense as it's used once ^_^

File: internal/clients/http/ec2/ec2_client.go
Comment (lzap) [inline internal/clients/http/ec2/ec2_client.go:176]:
```suggestion
func (c *ec2Client) FindNameByFingerprint(ctx context.Context, key *models.Pubkey) (string, error) {
```

Perhaps? We should probably only pass fingerprint, not the whole pubkey as well.

File: internal/jobs/launch_instance_aws.go
Comment (lzap) [inline internal/jobs/launch_instance_aws.go:79]: Isn't more straightforward to simply fetch this before the job and pass both pubkey and its resources? I mean, DB lookups are fine to be done outside of background job, they are fast. It's AWS API calls what should be performed in a job.

File: internal/jobs/pubkey_upload_aws_job.go
Comment (lzap) [reply internal/jobs/pubkey_upload_aws_job.go:91]: When I designed this, I was more focused on the job scheduler part than on the AWS part, so I sort of mindlessly created this "upload pubkey" job. I think it is the time to revisit this from the ground up and rethink this. Few points:

* The "aws handle" field does not have to be the "identifier" we use for keys, it can serve as an informative field too. I very much like the idea of searching pubkeys by fingerprint because it looks like on AWS this is what the API supports and it also enforces uniqueness therefore it is guaranteed to find 0 or 1 entry.
* The job is poorly named, I think we should think about our job scheduler as a small "ansible", it should be idempotent. Therefore I think we should rename it to EnsurePubkeyAWS or something like that - it should do whatever must be done to have that pubkey on the AWS account.
* It is absolutely no problem to perform more AWS API requests in a background job, we have all the time in the world, we can slow down jobs if we hit any API limits. So whoever want's to redesign this, they can even simply just query AWS first, if a key is found, then we can compare our handle and update it (or maybe it can be completely dropped we don't know if this is possible on other clouds). If it's not found, upload it and update our records.
* I think it is a good time to think about documenting each job, I suggest to have something close to the code, so code comment describing what the job should do would be a really nice idea. We can maybe start writing the new "job story" on the next tech meeting and once we agree we have a good approach that encapsulates all the edge cases, we can go ahead and implement it.

File: internal/jobs/launch_instance_aws.go
Comment (ezr-ondrej) [reply internal/jobs/launch_instance_aws.go:75]: I guess, but what do you think by passed to the job? the whole model in args? thinking that feels unsafe to me to marshal and unmarshal it for no reason - it feels better to just fetch it from the DB in the job shrug

File: internal/jobs/launch_instance_aws.go
Comment (ezr-ondrej) [inline internal/jobs/launch_instance_aws.go:93]: so the Name should be now saved to PubkeyNameAWS and passed to this method from there.
It feels like workaround and I belive it would be great if we have job outputs, but Happy to discuss later and use workaround for now.

File: internal/dao/tests/main_test.go
Comment (ezr-ondrej) [inline internal/dao/tests/main_test.go:30]: I don't know if ppl had their config in this folder, but this makes bit more sense to me :thinking:

File: internal/jobs/common.go
Comment (lzap) [reply internal/jobs/common.go:13]: Nice find, I am not sure how generics could be helpful here, but go ahead and make a draft PR if you have something on your mind.

File: internal/clients/http/ec2/ec2_client.go
Comment (lzap) [inline internal/clients/http/ec2/ec2_client.go:201]: I am not sure if we should treat the 404 like a hard error, at most this should be warning, if ever.

Other than that, looks good.

File: internal/clients/stubs/ec2_stub.go
Comment (lzap) [inline internal/clients/stubs/ec2_stub.go:52]: Highly encourage to use proper data than fake one :-) Like proper fingerprint of a real key... :-)

File: internal/dao/pgx/reservation_pgx.go
Comment (lzap) [inline internal/dao/pgx/reservation_pgx.go:224]: We should be generally trying to remove all "unscoped" DAO functions in the future, it's a bit more work to join tables to get an account, but it is dramatically safer if we misuse this in the future. I am okay to merge this in, unless you want to do an extra mile.

File: internal/jobs/queue/dejq/dejq_queue.go
Comment (lzap) [inline internal/jobs/queue/dejq/dejq_queue.go:33]: Can you also change the queue name? We do not persist queues yet so you can do this cheaply.

File: internal/models/reservation_model.go
Comment (lzap) [inline internal/models/reservation_model.go:64]: I suggest to be consistent:

```suggestion
	PubkeyName string `json:"pubkey_name"`
```

File: internal/jobs/launch_instance_aws.go
Comment (lzap) [reply internal/jobs/launch_instance_aws.go:93]: Not sure what you mean, this looks okay to me, like a stable and maintainable code. I don't think we are getting job pipeline anytime soon, even dbjobqueue does not support it (it is stored in DB yes but dependant jobs cannot access it I think).

File: internal/dao/tests/main_test.go
Comment (ezr-ondrej) [reply internal/dao/tests/main_test.go:30]: I do not want to start from weird directory, I want to run `make integration-tests` from the root directory of the project. I think it's sensible. Without this lookup it doesn't work. It's not `CWD` in this case, it's the main fail location - which is this file, so it is relative path to this file.

My aim was to make it work with config in the main `config/` directory, but I might have done something wrong :)

File: internal/clients/http/ec2/ec2_client.go
Comment (ezr-ondrej) [reply internal/clients/http/ec2/ec2_client.go:201]: well the error gets caught in the only caller, but if you think we should just return empty string, I'm ok with it, but feels bit weird, so I thought error is bit cleaner there.

File: internal/clients/stubs/ec2_stub.go
Comment (ezr-ondrej) [reply internal/clients/stubs/ec2_stub.go:52]: Well it is a special value at this moment, so I wanted to make it extra clear it is a special value. Real value might confuse ppl where it is defined. With real data, I'd want the key to actually be in the mock (added).

But that might be cleanest approach, I was just lazy to define fixtures for ex2, I needed to write a lot of setup code for tests already :) but if we want that, I'm happy to work on it.

File: internal/dao/pgx/reservation_pgx.go
Comment (ezr-ondrej) [reply internal/dao/pgx/reservation_pgx.go:224]: Well I don't want to. At least not without discussion, as complicating the queries might have some performance impact, I'd love to discuss if we don't want to do it in two queries. Which makes it a problem on its own IMHO and would slow down this PR which is already quite big and will be hard to get in :)

File: internal/jobs/launch_instance_aws.go
Comment (ezr-ondrej) [reply internal/jobs/launch_instance_aws.go:93]: right :) well workaround because we fetch record from database just to get output from previsous job (and the "column" exists just to be passed from one job to the other ;))

File: internal/dao/tests/main_test.go
Comment (ezr-ondrej) [reply internal/dao/tests/main_test.go:30]: I thought it will be no brainer, but if it needs to be discussed, lets take it to another PR: #337.

File: internal/clients/http/ec2/ec2_client.go
Comment (ezr-ondrej) [reply internal/clients/http/ec2/ec2_client.go:201]: Happy to indicate not found in other means, but would you have a suggestion?

File: internal/jobs/queue/dejq/dejq_queue.go
Comment (ezr-ondrej) [reply internal/jobs/queue/dejq/dejq_queue.go]: Done

File: internal/models/reservation_model.go
Comment (ezr-ondrej) [reply internal/models/reservation_model.go]: Done

File: internal/clients/http/ec2/ec2_client.go
Comment (lzap) [reply internal/clients/http/ec2/ec2_client.go:201]: Nah you are right that the caller should be responsible for skipping errors. This is fine.

File: internal/clients/stubs/ec2_stub.go
Comment (lzap) [reply internal/clients/stubs/ec2_stub.go:52]: I do not understand why you want to use this "special" value. Can't you simply update the stub to actually accept real Pubkey entries (with real data) so you can then get pubkey by name using it's real fingerprint? We already have several real pubkeys in the test suite (mine, Anna's...) and there is nothing wrong with creating stubs that actually do stuff. I actually thing although it will take more time, it will pay off as soon as it gets more complex later on.

File: internal/dao/pgx/reservation_pgx.go
Comment (lzap) [reply internal/dao/pgx/reservation_pgx.go:224]: You want to discuss SQL JOIN? Okay, but prepare good arguments for "joins are slow" in this case. My arguments will be "it's security, let's be extra careful". We already do this in several functions, I remember writing this myself.

Comment (lzap) [reply internal/dao/pgx/reservation_pgx.go:224]: Given the function prerix, I do not want to block this PR on this, but generally I would like us not to have any Unscoped functions, I do not see big reason not to join accounts table on a query that returns a _single row_.

**Comment by lzap** [comment]
Other than that, looks good.

File: internal/clients/stubs/ec2_stub.go
Comment (lzap) [inline internal/clients/stubs/ec2_stub.go:19]: Heh, what this is used for? Why don't you simply use `len(Imported)`?

File: internal/clients/stubs/ec2_stub.go
Comment (lzap) [inline internal/clients/stubs/ec2_stub.go:49]: This should really be `len(mock.Imported)+1`

File: internal/clients/stubs/ec2_stub.go
Comment (lzap) [inline internal/clients/stubs/ec2_stub.go:68]: Thank you this is WAAAY better than the old proposal.

**Comment by lzap** [comment]
Looks great, please explain why you don't use `len` and we can merge (linter issue).

File: internal/clients/stubs/ec2_stub.go
Comment (ezr-ondrej) [reply internal/clients/stubs/ec2_stub.go:19]: because for some reason it didn't occure to me xD Good idea :+1:

File: internal/dao/stubs/reservation_dao.go
Comment (lzap) [inline internal/dao/stubs/reservation_dao.go:12]: Oh my god we have it everywhere, it was a copy paste :-)

Ok not here, will fix this.

File: internal/jobs/launch_instance_aws.go
Comment (lzap) [reply internal/jobs/launch_instance_aws.go:75]: Right.

**Review by lzap — CHANGES_REQUESTED** [review]
The code looks good, but there is a problem in the workflow.

* PKR is looked up, let's assume it already exists in the db.
* Then presence of key is checked on AWS, assume it was not found (user deleted it).
* Therefore tag (of the existing pkr) is randomized and it is imported.
* Ssh key name is stored in the reservation.
* Finally, and here is the problem, since pkr is an existing record, ID != 0, therefore the newly generated tag will not be stored.

I think the solution is to remove `pkr.Tag = ""` and just call `RandomizeTag()` which will only randomize when pkr is not existing record.

When on it, please update the "Uploading/uploaded public key" status on line 52/53 to something like "Ensuring/Ensured" maybe. Not sure about it, we can perhaps keep it.

**Comment by lzap** [comment]
Also if both pkr and ssh-key exist but the tag/handle do not match, the pkr record is not updated. I think we should do this, it's one extra if statement (compare those two values and do update if they don't match AWS being the source of "truth"). I know we will eventually drop pkr but what if not for some reason - we want to have accurate data.

**Pull request merged**
